internal/discovery/hazelcast_cloud_discovery.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"net/http"
+
+	"encoding/json"
+
+	"io/ioutil"
+
+	"strings"
+
+	"time"
+
+	"github.com/hazelcast/hazelcast-go-client/core"
+	"github.com/hazelcast/hazelcast-go-client/internal/iputil"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+)
+
+const (
+	cloudURL          = "https://coordinator.hazelcast.cloud/cluster/discovery?token="
+	privateAddressStr = "private-address"
+	publicAddressStr  = "public-address"
+)
+
+type nodeDiscoverer func() (map[string]core.Address, error)
+
+type HazelcastCloud struct {
+	endPointURL               string
+	connectionTimeoutInMillis int64
+	discoverNodes             nodeDiscoverer
+}
+
+func NewHazelcastCloud(cloudToken string, connectionTimeoutInMillis int64) *HazelcastCloud {
+	hzCloud := &HazelcastCloud{}
+	hzCloud.endPointURL = cloudURL + cloudToken
+	hzCloud.connectionTimeoutInMillis = connectionTimeoutInMillis
+	hzCloud.discoverNodes = hzCloud.discoverNodesInternal
```

Comment:
nodeDiscoverer type and this line looks confusing. Why do we need that ? 
Are we trying to avoid a public method ? `DiscoverNodes`

Reply:
Hmm. It looks like it to do mock testing
https://github.com/hazelcast/hazelcast-go-client/pull/279/files#diff-64b25b464ccecbf58536edcc6ed4ffb8R41